From: Andreas Gruenbacher <agruen@suse.de>
Subject: Disable NFSv4 deamon-side POSIX ACL <=> NFSv4 ACL mapping
References: 169033

The NFSv4 deamon maps from local POSIX acls to NFSv4 acls on the wire,
which the NFSv4 client translates into "system.nfsv4_acl" extended
attributes, and vice versa.  This approach leads to errors when trying
to copy files from a NFSv4 mount to a different file system that does
not understand "system.nfsv4_acl" extended attributes (i.e., anything
but NFSv4).

Trying to do this mapping in the kernel is broken, and makes no sense
in my eyes.  There are two ways in which this could be fixed:

(1) Map NFSv4 acls to POSIX acls in the NFSv4 client, and map requests
    for setting POSIX acls into requests for setting NFSv4 acls, so
    that the pseudo-NFSv4 acls are never exposed to user space.
    
    The huge disadvantage of this approach is that as soon as the server
    or client supports "real" NFSv4 ACLs that do not match the limited
    pattern that the mapping understands, things will turn very ugly.

(2) Remove the mapping code entirely, and implement the NFSACL sideband
    protocol for NFSv4 as well. (Solaris seems to do this.)  This will
    not block us from implementing full NFSv4 ACLs somewhen in the
    future. The NFSACL protocol is implemented for NFSv2 and NFSv3
    already, so adding NFSv4 support should not be be very hard.


This patch disables the POSIX ACL to NFSv4 ACL mapping in nfsd.

Because the POSIX ACL to NFSv4 ACL mapping found its way into the
upstream kernel, vanilla kernels will currently report that they
support NFSv4 ACLs, while they really only support POSIX ACLs. To
avoid the interoperability problems this would cause, also turn off 
NFSv4 ACL support in the client.


Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

---
 fs/nfs/nfs4proc.c         |    4 ++++
 include/linux/nfsd/nfsd.h |    6 +++---
 2 files changed, 7 insertions(+), 3 deletions(-)

--- linux-2.6.18.orig/fs/nfs/nfs4proc.c
+++ linux-2.6.18/fs/nfs/nfs4proc.c
@@ -2577,9 +2577,13 @@ int nfs4_proc_renew(struct nfs4_client *
 
 static inline int nfs4_server_supports_acls(struct nfs_server *server)
 {
+#if 0
 	return (server->caps & NFS_CAP_ACLS)
 		&& (server->acl_bitmask & ACL4_SUPPORT_ALLOW_ACL)
 		&& (server->acl_bitmask & ACL4_SUPPORT_DENY_ACL);
+#else
+	return 0;
+#endif
 }
 
 /* Assuming that XATTR_SIZE_MAX is a multiple of PAGE_CACHE_SIZE, and that
--- linux-2.6.18.orig/include/linux/nfsd/nfsd.h
+++ linux-2.6.18/include/linux/nfsd/nfsd.h
@@ -299,12 +299,12 @@ static inline int is_fsid(struct svc_fh 
  | FATTR4_WORD0_CHANGE          | FATTR4_WORD0_SIZE         | FATTR4_WORD0_LINK_SUPPORT     \
  | FATTR4_WORD0_SYMLINK_SUPPORT | FATTR4_WORD0_NAMED_ATTR   | FATTR4_WORD0_FSID             \
  | FATTR4_WORD0_UNIQUE_HANDLES  | FATTR4_WORD0_LEASE_TIME   | FATTR4_WORD0_RDATTR_ERROR     \
- | FATTR4_WORD0_ACLSUPPORT      | FATTR4_WORD0_CANSETTIME   | FATTR4_WORD0_CASE_INSENSITIVE \
+/*| FATTR4_WORD0_ACLSUPPORT*/   | FATTR4_WORD0_CANSETTIME   | FATTR4_WORD0_CASE_INSENSITIVE \
  | FATTR4_WORD0_CASE_PRESERVING | FATTR4_WORD0_CHOWN_RESTRICTED                             \
  | FATTR4_WORD0_FILEHANDLE      | FATTR4_WORD0_FILEID       | FATTR4_WORD0_FILES_AVAIL      \
  | FATTR4_WORD0_FILES_FREE      | FATTR4_WORD0_FILES_TOTAL  | FATTR4_WORD0_HOMOGENEOUS      \
  | FATTR4_WORD0_MAXFILESIZE     | FATTR4_WORD0_MAXLINK      | FATTR4_WORD0_MAXNAME          \
- | FATTR4_WORD0_MAXREAD         | FATTR4_WORD0_MAXWRITE     | FATTR4_WORD0_ACL)
+ | FATTR4_WORD0_MAXREAD         | FATTR4_WORD0_MAXWRITE   /*| FATTR4_WORD0_ACL*/)
 
 #define NFSD_SUPPORTED_ATTRS_WORD1                                                          \
 (FATTR4_WORD1_MODE              | FATTR4_WORD1_NO_TRUNC     | FATTR4_WORD1_NUMLINKS         \
@@ -320,7 +320,7 @@ static inline int is_fsid(struct svc_fh 
 
 /* These are the only attrs allowed in CREATE/OPEN/SETATTR. */
 #define NFSD_WRITEABLE_ATTRS_WORD0                                                          \
-(FATTR4_WORD0_SIZE              | FATTR4_WORD0_ACL                                         )
+(FATTR4_WORD0_SIZE             /*| FATTR4_WORD0_ACL*/                                       )
 #define NFSD_WRITEABLE_ATTRS_WORD1                                                          \
 (FATTR4_WORD1_MODE              | FATTR4_WORD1_OWNER         | FATTR4_WORD1_OWNER_GROUP     \
  | FATTR4_WORD1_TIME_ACCESS_SET | FATTR4_WORD1_TIME_METADATA | FATTR4_WORD1_TIME_MODIFY_SET)
